Allow users to bundle node_modules

lucasfcosta opened the pull request:
> Please notice that I had to rebase this PR on top of #576 so that we can actually parse streaming responses from Kibana.


# Summary

This PR allows users to bundle modules into the scripts they push.

I've marked it as draft as it still needs E2E tests.


# How to test this PR

To test this PR, try including a JS library into your journey, by importing it into the journey file.

Then push your journey with the extra imported module to Kibana.

You should be able to run the Journey just fine.

apmmachine commented:
## :broken_heart: Tests Failed


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-rum%2Felastic-synthetics%2FPR-580/detail/PR-580/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-rum%2Felastic-synthetics%2FPR-580/detail/PR-580/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-rum%2Felastic-synthetics%2FPR-580/detail/PR-580/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-rum%2Felastic-synthetics%2FPR-580/detail/PR-580/1//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://synthetics_580.docs-preview.app.elstc.co/diff)  [![preview](https://img.shields.io/badge/elastic-observability-blue)](https://ci-stats.elastic.co/app/apm/services/apm-ci/transactions/view?rangeFrom=2022-08-17T17:12:46.512Z&rangeTo=2022-08-17T17:32:46.512Z&transactionName=BUILD+apm-agent-rum%2Felastic-synthetics%2FPR-%7Bnumber%7D&transactionType=job&latencyAggregationType=avg&traceId=3feb3f437a59c6d2a37eaf2b3d5db4c1&transactionId=72b47845c6985cc2)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats




* Start Time: 2022-08-17T17:22:46.512+0000


* Duration: 8 min 58 sec




#### Test stats :test_tube:

| Test         | Results                         |
| ------------ | :-----------------------------: |
| Failed       | 6  |
| Passed       | 173  |
| Skipped      | 2 |
| Total        | 181   |


</p>
</details>

<!-- TEST RESULTS IF ANY-->

  ### Test errors [![6](https://img.shields.io/badge/6%20-red)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-rum%2Felastic-synthetics%2FPR-580/detail/PR-580/1//tests)
<details><summary>Expand to view the tests failures</summary><p>
    
    
##### `Test / Bundler build journey – Bundler build journey`


<ul>
<details><summary>Expand to view the error details</summary><p>

```
 null 
 ```
</p></details>
<details><summary>Expand to view the stacktrace</summary><p>

```
 Error: Build failed with 1 error:
asset:/var/lib/jenkins/workspace/nt-rum_elastic-synthetics_PR-580/src/github.com/elastic/synthetics/__tests__/e2e/uptime.journey.ts:27:18: ERROR: Could not resolve "axios"
    at failureErrorWithLog (/var/lib/jenkins/workspace/nt-rum_elastic-synthetics_PR-580/src/github.com/elastic/synthetics/node_modules/esbuild/lib/main.js:1599:15)
    at /var/lib/jenkins/workspace/nt-rum_elastic-synthetics_PR-580/src/github.com/elastic/synthetics/node_modules/esbuild/lib/main.js:1245:28
    at runOnEndCallbacks (/var/lib/jenkins/workspace/nt-rum_elastic-synthetics_PR-580/src/github.com/elastic/synthetics/node_modules/esbuild/lib/main.js:1030:63)
    at buildResponseToResult (/var/lib/jenkins/workspace/nt-rum_elastic-synthetics_PR-580/src/github.com/elastic/synthetics/node_modules/esbuild/lib/main.js:1243:7)
    at /var/lib/jenkins/workspace/nt-rum_elastic-synthetics_PR-580/src/github.com/elastic/synthetics/node_modules/esbuild/lib/main.js:1352:14
    at /var/lib/jenkins/workspace/nt-rum_elastic-synthetics_PR-580/src/github.com/elastic/synthetics/node_modules/esbuild/lib/main.js:662:9
    at handleIncomingPacket (/var/lib/jenkins/workspace/nt-rum_elastic-synthetics_PR-580/src/github.com/elastic/synthetics/node_modules/esbuild/lib/main.js:759:9)
    at Socket.readFromStdout (/var/lib/jenkins/workspace/nt-rum_elastic-synthetics_PR-580/src/github.com/elastic/synthetics/node_modules/esbuild/lib/main.js:629:7)
    at Socket.emit (node:events:527:28)
    at addChunk (node:internal/streams/readable:315:12)
    at readableAddChunk (node:internal/streams/readable:289:9)
    at Socket.Readable.push (node:internal/streams/readable:228:10)
    at Pipe.onStreamRead (node:internal/stream_base_commons:190:23) 
 ```
</p></details>
</ul>
    
##### `Test / Monitors api schema – Monitors api schema`


<ul>
<details><summary>Expand to view the error details</summary><p>

```
 null 
 ```
</p></details>
<details><summary>Expand to view the stacktrace</summary><p>

```
 Error: Build failed with 1 error:
asset:/var/lib/jenkins/workspace/nt-rum_elastic-synthetics_PR-580/src/github.com/elastic/synthetics/__tests__/fixtures/example.journey.ts:26:30: ERROR: [plugin: esbuild-multiasset-plugin] Build failed with 1 error:
node_modules/playwright-core/lib/server/electron/electron.js:172:51: ERROR: Could not resolve "electron/index.js"
    at failureErrorWithLog (/var/lib/jenkins/workspace/nt-rum_elastic-synthetics_PR-580/src/github.com/elastic/synthetics/node_modules/esbuild/lib/main.js:1599:15)
    at /var/lib/jenkins/workspace/nt-rum_elastic-synthetics_PR-580/src/github.com/elastic/synthetics/node_modules/esbuild/lib/main.js:1245:28
    at runOnEndCallbacks (/var/lib/jenkins/workspace/nt-rum_elastic-synthetics_PR-580/src/github.com/elastic/synthetics/node_modules/esbuild/lib/main.js:1030:63)
    at buildResponseToResult (/var/lib/jenkins/workspace/nt-rum_elastic-synthetics_PR-580/src/github.com/elastic/synthetics/node_modules/esbuild/lib/main.js:1243:7)
    at /var/lib/jenkins/workspace/nt-rum_elastic-synthetics_PR-580/src/github.com/elastic/synthetics/node_modules/esbuild/lib/main.js:1352:14
    at /var/lib/jenkins/workspace/nt-rum_elastic-synthetics_PR-580/src/github.com/elastic/synthetics/node_modules/esbuild/lib/main.js:662:9
    at handleIncomingPacket (/var/lib/jenkins/workspace/nt-rum_elastic-synthetics_PR-580/src/github.com/elastic/synthetics/node_modules/esbuild/lib/main.js:759:9)
    at Socket.readFromStdout (/var/lib/jenkins/workspace/nt-rum_elastic-synthetics_PR-580/src/github.com/elastic/synthetics/node_modules/esbuild/lib/main.js:629:7)
    at Socket.emit (node:events:527:28)
    at addChunk (node:internal/streams/readable:315:12) 
 ```
</p></details>
</ul>
    
##### `Test / Monitors build monitor schema monitor – Monitors build monitor schema monitor`


<ul>
<details><summary>Expand to view the error details</summary><p>

```
 null 
 ```
</p></details>
<details><summary>Expand to view the stacktrace</summary><p>

```
 Error: Build failed with 1 error:
asset:/var/lib/jenkins/workspace/nt-rum_elastic-synthetics_PR-580/src/github.com/elastic/synthetics/__tests__/fixtures/example.journey.ts:26:30: ERROR: [plugin: esbuild-multiasset-plugin] Build failed with 1 error:
node_modules/playwright-core/lib/server/electron/electron.js:172:51: ERROR: Could not resolve "electron/index.js"
    at failureErrorWithLog (/var/lib/jenkins/workspace/nt-rum_elastic-synthetics_PR-580/src/github.com/elastic/synthetics/node_modules/esbuild/lib/main.js:1599:15)
    at /var/lib/jenkins/workspace/nt-rum_elastic-synthetics_PR-580/src/github.com/elastic/synthetics/node_modules/esbuild/lib/main.js:1245:28
    at runOnEndCallbacks (/var/lib/jenkins/workspace/nt-rum_elastic-synthetics_PR-580/src/github.com/elastic/synthetics/node_modules/esbuild/lib/main.js:1030:63)
    at buildResponseToResult (/var/lib/jenkins/workspace/nt-rum_elastic-synthetics_PR-580/src/github.com/elastic/synthetics/node_modules/esbuild/lib/main.js:1243:7)
    at /var/lib/jenkins/workspace/nt-rum_elastic-synthetics_PR-580/src/github.com/elastic/synthetics/node_modules/esbuild/lib/main.js:1352:14
    at /var/lib/jenkins/workspace/nt-rum_elastic-synthetics_PR-580/src/github.com/elastic/synthetics/node_modules/esbuild/lib/main.js:662:9
    at handleIncomingPacket (/var/lib/jenkins/workspace/nt-rum_elastic-synthetics_PR-580/src/github.com/elastic/synthetics/node_modules/esbuild/lib/main.js:759:9)
    at Socket.readFromStdout (/var/lib/jenkins/workspace/nt-rum_elastic-synthetics_PR-580/src/github.com/elastic/synthetics/node_modules/esbuild/lib/main.js:629:7)
    at Socket.emit (node:events:527:28)
    at addChunk (node:internal/streams/readable:315:12) 
 ```
</p></details>
</ul>
    
##### `Test / Plugin bundle jouneys – Plugin bundle jouneys`


<ul>
<details><summary>Expand to view the error details</summary><p>

```
 null 
 ```
</p></details>
<details><summary>Expand to view the stacktrace</summary><p>

```
 Error: Build failed with 1 error:
asset:/var/lib/jenkins/workspace/nt-rum_elastic-synthetics_PR-580/src/github.com/elastic/synthetics/__tests__/e2e/uptime.journey.ts:27:18: ERROR: Could not resolve "axios"
    at failureErrorWithLog (/var/lib/jenkins/workspace/nt-rum_elastic-synthetics_PR-580/src/github.com/elastic/synthetics/node_modules/esbuild/lib/main.js:1599:15)
    at /var/lib/jenkins/workspace/nt-rum_elastic-synthetics_PR-580/src/github.com/elastic/synthetics/node_modules/esbuild/lib/main.js:1245:28
    at runOnEndCallbacks (/var/lib/jenkins/workspace/nt-rum_elastic-synthetics_PR-580/src/github.com/elastic/synthetics/node_modules/esbuild/lib/main.js:1030:63)
    at buildResponseToResult (/var/lib/jenkins/workspace/nt-rum_elastic-synthetics_PR-580/src/github.com/elastic/synthetics/node_modules/esbuild/lib/main.js:1243:7)
    at /var/lib/jenkins/workspace/nt-rum_elastic-synthetics_PR-580/src/github.com/elastic/synthetics/node_modules/esbuild/lib/main.js:1352:14
    at /var/lib/jenkins/workspace/nt-rum_elastic-synthetics_PR-580/src/github.com/elastic/synthetics/node_modules/esbuild/lib/main.js:662:9
    at handleIncomingPacket (/var/lib/jenkins/workspace/nt-rum_elastic-synthetics_PR-580/src/github.com/elastic/synthetics/node_modules/esbuild/lib/main.js:759:9)
    at Socket.readFromStdout (/var/lib/jenkins/workspace/nt-rum_elastic-synthetics_PR-580/src/github.com/elastic/synthetics/node_modules/esbuild/lib/main.js:629:7)
    at Socket.emit (node:events:527:28)
    at addChunk (node:internal/streams/readable:315:12) 
 ```
</p></details>
</ul>
    
##### `Test / Push API handle streamed response – Push API handle streamed response`


<ul>
<details><summary>Expand to view the error details</summary><p>

```
 null 
 ```
</p></details>
<details><summary>Expand to view the stacktrace</summary><p>

```
 Error: expect(received).toMatchSnapshot()

Snapshot name: `Push API handle streamed response 1`

- Snapshot  -  5
+ Received  + 25

  > preparing all monitors
- > creating all monitors
- > chunk 1
- > chunk 2
- > chunk 3
- ✓ Pushed
+ Error: Build failed with 1 error:
+ asset:/var/lib/jenkins/workspace/nt-rum_elastic-synthetics_PR-580/src/github.com/elastic/synthetics/__tests__/push/new-project/test.journey.ts:1:31: ERROR: [plugin: esbuild-multiasset-plugin] Build failed with 1 error:
+ ../../../node_modules/playwright-core/lib/server/electron/electron.js:172:51: ERROR: Could not resolve "electron/index.js"
+     at failureErrorWithLog (/var/lib/jenkins/workspace/nt-rum_elastic-synthetics_PR-580/src/github.com/elastic/synthetics/node_modules/esbuild/lib/main.js:1599:15)
+     at /var/lib/jenkins/workspace/nt-rum_elastic-synthetics_PR-580/src/github.com/elastic/synthetics/node_modules/esbuild/lib/main.js:1245:28
+     at runOnEndCallbacks (/var/lib/jenkins/workspace/nt-rum_elastic-synthetics_PR-580/src/github.com/elastic/synthetics/node_modules/esbuild/lib/main.js:1030:63)
+     at buildResponseToResult (/var/lib/jenkins/workspace/nt-rum_elastic-synthetics_PR-580/src/github.com/elastic/synthetics/node_modules/esbuild/lib/main.js:1243:7)
+     at /var/lib/jenkins/workspace/nt-rum_elastic-synthetics_PR-580/src/github.com/elastic/synthetics/node_modules/esbuild/lib/main.js:1352:14
+     at /var/lib/jenkins/workspace/nt-rum_elastic-synthetics_PR-580/src/github.com/elastic/synthetics/node_modules/esbuild/lib/main.js:662:9
+     at handleIncomingPacket (/var/lib/jenkins/workspace/nt-rum_elastic-synthetics_PR-580/src/github.com/elastic/synthetics/node_modules/esbuild/lib/main.js:759:9)
+     at Socket.readFromStdout (/var/lib/jenkins/workspace/nt-rum_elastic-synthetics_PR-580/src/github.com/elastic/synthetics/node_modules/esbuild/lib/main.js:629:7)
+     at Socket.emit (node:events:527:28)
+     at Socket.emit (node:domain:475:12) {
+   errors: [
+     {
+       detail: [Error],
+       location: [Object],
+       notes: [Array],
+       pluginName: 'esbuild-multiasset-plugin',
+       text: 'Build failed with 1 error:\n' +
+         '../../../node_modules/playwright-core/lib/server/electron/electron.js:172:51: ERROR: Could not resolve "electron/index.js"'
+     }
+   ],
+   warnings: []
+ }
  ↵
    at Object.<anonymous> (/var/lib/jenkins/workspace/nt-rum_elastic-synthetics_PR-580/src/github.com/elastic/synthetics/__tests__/push/index.test.ts:217:22)
    at processTicksAndRejections (node:internal/process/task_queues:96:5) 
 ```
</p></details>
</ul>
    
##### `Test / Push API handle sync response – Push API handle sync response`


<ul>
<details><summary>Expand to view the error details</summary><p>

```
 null 
 ```
</p></details>
<details><summary>Expand to view the stacktrace</summary><p>

```
 Error: expect(received).toMatchSnapshot()

Snapshot name: `Push API handle sync response 1`

- Snapshot  -  2
+ Received  + 25

  > preparing all monitors
- > creating all monitors
- ✓ Pushed
+ Error: Build failed with 1 error:
+ asset:/var/lib/jenkins/workspace/nt-rum_elastic-synthetics_PR-580/src/github.com/elastic/synthetics/__tests__/push/new-project/test.journey.ts:1:31: ERROR: [plugin: esbuild-multiasset-plugin] Build failed with 1 error:
+ ../../../node_modules/playwright-core/lib/server/electron/electron.js:172:51: ERROR: Could not resolve "electron/index.js"
+     at failureErrorWithLog (/var/lib/jenkins/workspace/nt-rum_elastic-synthetics_PR-580/src/github.com/elastic/synthetics/node_modules/esbuild/lib/main.js:1599:15)
+     at /var/lib/jenkins/workspace/nt-rum_elastic-synthetics_PR-580/src/github.com/elastic/synthetics/node_modules/esbuild/lib/main.js:1245:28
+     at runOnEndCallbacks (/var/lib/jenkins/workspace/nt-rum_elastic-synthetics_PR-580/src/github.com/elastic/synthetics/node_modules/esbuild/lib/main.js:1030:63)
+     at buildResponseToResult (/var/lib/jenkins/workspace/nt-rum_elastic-synthetics_PR-580/src/github.com/elastic/synthetics/node_modules/esbuild/lib/main.js:1243:7)
+     at /var/lib/jenkins/workspace/nt-rum_elastic-synthetics_PR-580/src/github.com/elastic/synthetics/node_modules/esbuild/lib/main.js:1352:14
+     at /var/lib/jenkins/workspace/nt-rum_elastic-synthetics_PR-580/src/github.com/elastic/synthetics/node_modules/esbuild/lib/main.js:662:9
+     at handleIncomingPacket (/var/lib/jenkins/workspace/nt-rum_elastic-synthetics_PR-580/src/github.com/elastic/synthetics/node_modules/esbuild/lib/main.js:759:9)
+     at Socket.readFromStdout (/var/lib/jenkins/workspace/nt-rum_elastic-synthetics_PR-580/src/github.com/elastic/synthetics/node_modules/esbuild/lib/main.js:629:7)
+     at Socket.emit (node:events:527:28)
+     at Socket.emit (node:domain:475:12) {
+   errors: [
+     {
+       detail: [Error],
+       location: [Object],
+       notes: [Array],
+       pluginName: 'esbuild-multiasset-plugin',
+       text: 'Build failed with 1 error:\n' +
+         '../../../node_modules/playwright-core/lib/server/electron/electron.js:172:51: ERROR: Could not resolve "electron/index.js"'
+     }
+   ],
+   warnings: []
+ }
  ↵
    at Object.<anonymous> (/var/lib/jenkins/workspace/nt-rum_elastic-synthetics_PR-580/src/github.com/elastic/synthetics/__tests__/push/index.test.ts:209:22)
    at processTicksAndRejections (node:internal/process/task_queues:96:5) 
 ```
</p></details>
</ul>
    
</p></details>


<!-- STEPS ERRORS IF ANY -->














### Steps errors [![1](https://img.shields.io/badge/1%20-red)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-rum%2Felastic-synthetics%2FPR-580/detail/PR-580/1//pipeline)
  <details><summary>Expand to view the steps failures</summary>
  <p>


  
  
  
  
##### `Runs the tests`
<ul>
<li>Took 2 min 59 sec . View more details <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-rum/pipelines/elastic-synthetics/pipelines/PR-580/runs/1/steps/341/log/?start=0">here</a></li>
<li>Description: <code>./node_modules/.bin/jest --ci --reporters=default --reporters=jest-junit</code></l1>
</ul>
  
  </p>
  </details>



## :robot: GitHub comments

To re-run your PR in the CI, just comment with:



- `/test` : Re-trigger the build.




<!--COMMENT_GENERATED_WITH_ID_comment.id-->

vigneshshanmugam commented on src/push/plugin.ts:
```diff
@@ -88,13 +88,14 @@ export function MultiAssetPlugin(callback: PluginCallback): esbuild.Plugin {
             [entryPath]: entryPath,
           },
-          bundle: false,
+          bundle: true,
```
this would change our journey files format IIRC.

lucasfcosta replied:
That's interesting, do you mean the file format itself (the extension)? Or do you mean the content format (i.e. bundling the CommonJS infra)?

In my tests, I added an `fs.writeFile` to the code so I could save the generated zips to my disk. Then, once I opened them, I did manage to see that the file extension was `.js` and that the content was also valid JavaScript.

I also checked the docs here: https://esbuild.github.io/api/#bundle and couldn't see a reference for format changes.

Can you go into a bit more detail about your concern here? Perhaps I'm missing something or misunderstood.

vigneshshanmugam replied:
I mean the content of the journey itself. Lets say when you push a journey file `file.journey.ts`, What does the content look like? Did it get changed to something else and whether the files were bundled - concatenated in to a single file for the files(utils, helpers) that were imported by the journey?

lucasfcosta commented:
CC @vigneshshanmugam feel free to merge this whenever you're happy with it, all tests are now passing 😄

vigneshshanmugam reviewed (changes requested):
Even though this solution works, this defeats the purpose of our Plugin and kind of bundles everything together in a single file. 

The goal of our push command is to replicate exactly how your folder structure is on your local and run the same inside HB when pushed. This would make it presentable in Kibana when we add unzipping support and allow users to view and edit monitors. 

TLDR: If i push the `advanced.journey.ts` to Kibana, the structure should like
```
// Main branch
/example/todos/advanced.journey.ts
/example/todos/helper.ts


// PR branch
/example/todos/advanced.journey.ts
```

lucasfcosta commented:
@vigneshshanmugam can you double check whether you had the last version during your test? That behaviour you mentioned was happening before the changes I did yesterday morning.

Right now, I just tried to push a journey which depends on another local file and it did separate the necessary files.

To test that, I added the following line to `src/push/bundler.js` in the `build` function to get the ZIP file:

```ts
  async build(entry: string, output: string) {
    await this.prepare(entry);
    await this.zip(output);
    const data = await this.encode(output);
    await writeFile('/tmp/new-synth.zip', data, { encoding: 'base64' });
    await this.cleanup(output);
    return data;
  }
```

Then I extracted the `new-synth` zip file and got two separate files:

<img width="789" alt="Screenshot 2022-08-24 at 13 19 45" src="https://user-images.githubusercontent.com/6868147/186417109-b8567791-dc99-46b0-a1e9-69d873647814.png">

Is this what you meant? Or did I misunderstand?

paulb-elastic commented:
As discussed in Tech Sync, we should ensure this feature includes a check and handling the failure of the compressed bundle being too big for either the Saved Object, or configmap to support, and feed this back to the user (e.g. _...failed to create the monitor as the bundle size is bigger than the maximum 800KB..._)
cc @lucasfcosta

lucasfcosta commented:
@vigneshshanmugam I was looking into not bundling the `node_modules` but in that case we'd have to implement the resolution ourselves as the plugin we use bundles them by default and it doesn't seem like we can change that.

Considering that Kibana only shows the code _inside_ the step, as shown in the screenshot below, what do you think of us merging this one as is?

![Screenshot 2022-08-31 at 10 35 28](https://user-images.githubusercontent.com/6868147/187648072-1e788a9d-3afa-49b5-b277-4a2eb02dbe58.png)

vigneshshanmugam commented on CHANGELOG.md:
```diff
@@ -5,15 +5,79 @@ All notable changes to this project will be documented in this file.
 **For detailed release notes, please refer to the [GitHub
 releases](https://github.com/elastic/synthetics/releases) page.**
 
+## v1.0.0-beta.32 (2022-08-24)
```
can you rebase @lucasfcosta ?

lucasfcosta replied:
Done!

vigneshshanmugam commented on src/push/plugin.ts:
```diff
@@ -61,21 +68,43 @@ export function MultiAssetPlugin(callback: PluginCallback): esbuild.Plugin {
+  // it may import its own local dependencies, so we must
+  // make sure those will be resolved using Node's resolution
+  // algorithm, as they're still "node_modules" that we must bundle
+  const isLocalSyntheticsModule = (str: string) => {
```
Why not do str.includes`@elastic/synthetics`, it would work with source node modules and synthetics itself? We would get absolute path all the time, so this should be good enough?

lucasfcosta replied:
I might have misunderstood your comment, but yes we do get the absolute path (when importing the `.js`, not always). That's the exact reason we can't use the `includes` you suggested because when importing the local file the alias will not be the name we're passing in.

Is this what you meant? I thought I might have misunderstood because even if what you suggest happens always that means we can never use the suggested `includes`.

StuGray86 commented:
Hi guys.

We're currently waiting on this change, is there any timeframe for its release?

vigneshshanmugam reviewed (approved):
LGTM 🎉

awahab07 commented:
**Post FF Testing**

I was able to import a simple library like `uuid` and successfully create, push and run the journey.

Script I used:        
```typescript
import { journey, step, monitor } from '@elastic/synthetics';

import { v4 as uuidv4 } from 'uuid';

journey('Journey uuid Test', ({ page, params }) => {
  monitor.use({
    id: 'uuid-lib-test-001',
    enabled: true,
    schedule: 20,
  });

  step('Generate UUID', async () => {
    const uuid = uuidv4();
    await page.goto('https://onlinetexttools.com/convert-text-to-image');
    const textAreaLocator = page.locator('textarea').first();
    await textAreaLocator.fill(uuid);
    await textAreaLocator.type(' ');
  });
});

```
<img width="1206" alt="image" src="https://user-images.githubusercontent.com/2748376/193883672-d881de02-d1c8-4def-a829-b5ec46734991.png">

However, I did find an issue with bundling certain imports and raised a [bug](https://github.com/elastic/synthetics/issues/619).
